build_runner/test/package_graph/package_graph_test.dart

```diff
@@ -150,6 +150,63 @@ void main() {
           throwsA(anything));
     });
   });
+
+  group('orderedPackages', () {
+    test('with two sub trees', () {
+      var a = new PackageNode('a', '1.0.0', PackageDependencyType.path, null);
+      var left1 =
+          new PackageNode('left1', '1.0.0', PackageDependencyType.pub, null);
+      var left2 =
+          new PackageNode('left2', '1.0.0', PackageDependencyType.pub, null);
+      var right1 =
+          new PackageNode('right1', '1.0.0', PackageDependencyType.pub, null);
+      var right2 =
+          new PackageNode('right2', '1.0.0', PackageDependencyType.pub, null);
+      a.dependencies.addAll([left1, right1]);
+      left1.dependencies.add(left2);
+      right1.dependencies.add(right2);
+      var graph = new PackageGraph.fromRoot(a);
+      var inOrder = graph.orderedPackages.map((n) => n.name).toList();
+      expect(inOrder, containsAllInOrder(['left2', 'left1', 'a']));
+      expect(inOrder, containsAllInOrder(['right2', 'right1', 'a']));
+    });
+
+    test('handles cycles', () {
+      var a = new PackageNode('a', '1.0.0', PackageDependencyType.path, null);
+      var b = new PackageNode('b', '1.0.0', PackageDependencyType.path, null);
+      a.dependencies.add(b);
+      b.dependencies.add(a);
+      var graph = new PackageGraph.fromRoot(a);
+      var inOrder = graph.orderedPackages.map((n) => n.name).toList();
+      expect(inOrder, containsAllInOrder(['a', 'b']));
```

Comment:
nit: technically the order here is non-deterministic so you could just use `contains`.

Also it would probably be good to add another test with a cycle that doesn't start from the root.

Reply:
This is a bug in my implementation. I want to guarantee that the root package is always last, but my implementation doesn't have that guarantee. The recursive solution does though so I kept this test, fixed the order, and updated the test name for clarity.

Also added a non-root cycle test